File: leetcode/8.string-to-integer-atoi.py
# @leet start
class Solution:
    def myAtoi(self, s: str) -> int:
        n = len(s)
        i = 0

        # Whitespace
        while i < n and s[i] == " ":
            i += 1

        # Sign
        sign = 1

        if i < n and s[i] in {"-", "+"}:
            if s[i] == "-":
                sign = -1

            i += 1

        start = i

        # Digits
        while i < n and s[i].isdigit():
            i += 1

        return min(max(sign * int(s[start:i]) if start < i else 0, -(2**31)), 2**31 - 1)


# @leet end

File: leetcode/8.string-to-integer-atoi.ts
// @leet start
function myAtoi(s: string): number {
  let start = 0;

  while (start < s.length && s[start] === " ") {
    start++;
  }

  let sign = 1;

  if (start < s.length && "+-".includes(s[start])) {
    if (s[start] === "-") {
      sign = -1;
    }

    start++;
  }

  let end = start;

  while (end < s.length && /\d/.test(s[end])) {
    end++;
  }

  return Math.min(
    Math.max(sign * Number(s.slice(start, end)), -(2 ** 31)),
    2 ** 31 - 1,
  );
}
// @leet end
